src/Faithlife.Tracing.Zipkin/W3CPropagation.cs

```diff
@@ -0,0 +1,161 @@
+			var traceIdLow = long.Parse(traceParent.Substring(19, 16), NumberStyles.AllowHexSpecifier, CultureInfo.InvariantCulture);
+			var spanId = long.Parse(traceParent.Substring(36, 16), NumberStyles.AllowHexSpecifier, CultureInfo.InvariantCulture);
+			var flags = byte.Parse(traceParent.Substring(53, 2), NumberStyles.AllowHexSpecifier, CultureInfo.InvariantCulture);
+			var isSampled = ((flags & 0x1) == 0x1) ? true : default(bool?);
+			return new W3CTraceContext(traceIdHigh, traceIdLow, spanId, parentSpanId, isSampled, isDebug, traceContext);
+		}
+
+		internal static bool IsValidTraceParent(string traceParent)
+		{
+			if (traceParent == null || traceParent.Length != 55)
+				return false;
+
+			if (traceParent[0] != '0' || traceParent[1] != '0' || traceParent[2] != '-')
+				return false;
+
+			for (int i = 3; i < traceParent.Length; i++)
+			{
+				var ch = traceParent[i];
+				if (i == 35 || i == 52)
+				{
+					if (ch != '-')
+						return false;
+				}
+				else
+				{
+					if (!((ch >= '0' && ch <= '9') || (ch >= 'a' && ch <= 'f') || (ch >= 'A' && ch <= 'F')))
+						return false;
+				}
+			}
+
+			return true;
+		}
+
+		private sealed class W3CInjector<TCarrier> : IInjector<TCarrier>
+		{
+			public W3CInjector(Setter<TCarrier, string> setter) => m_setter = setter;
+
+			public void Inject(ITraceContext traceContext, TCarrier carrier)
+			{
+				var (parent, context) = CreateHeaderValues(traceContext);
+				m_setter(carrier, c_traceParentHeader, parent);
+				m_setter(carrier, c_traceContextHeader, context);
+			}
+
+			readonly Setter<TCarrier, string> m_setter;
+		}
+
+		private sealed class W3CExtractor<TCarrier> : IExtractor<TCarrier>
+		{
+			public W3CExtractor(Getter<TCarrier, string> getter) => m_getter = getter;
+
+			public ITraceContext Extract(TCarrier carrier)
+			{
+				var traceParent = m_getter(carrier, c_traceParentHeader);
+				if (!IsValidTraceParent(traceParent))
+					return null;
+
+				var context = m_getter(carrier, c_traceContextHeader);
+				return W3CPropagation.Extract(traceParent, context);
+			}
+
+			readonly Getter<TCarrier, string> m_getter;
+		}
+
+		internal sealed class W3CTraceContext : ITraceContext
+		{
+			public W3CTraceContext(long traceIdHigh, long traceIdLow, long spanId, long? parentSpanId, bool? isSampled, bool isDebug, string context)
+			{
+				TraceIdHigh = traceIdHigh;
+				TraceId = traceIdLow;
+				SpanId = spanId;
+				ParentSpanId = parentSpanId;
+				Sampled = isSampled;
+				Debug = isDebug;
+				if (context != null)
+					Extra = new List<object> { context };
+			}
+
+			public long TraceIdHigh { get; }
+			public long TraceId { get; }
+			public long? ParentSpanId { get; }
+			public long SpanId { get; }
+			public bool? Sampled { get; }
+			public bool Debug { get; }
+			public List<object> Extra { get; }
+		}
+
+		const string c_vendorKey = "faithlifezipkin=";
```

Comment:
I might call this `c_vendorKeyEquals` or something; the trailing equals isn't implied by the name.